src/wrapper/objects/global_ref.rs

```diff
@@ -1,9 +1,13 @@
-use std::{convert::From, sync::Arc};
+use std::{mem, ops::Deref, sync::Arc};
 
 use log::{debug, warn};
 
 use crate::{errors::Result, objects::JObject, sys, JNIEnv, JavaVM};
 
+// Note: `GlobalRef` must not implement `Into<JObject>`! If it did, then it would be possible to
+// wrap it in `AutoLocal`, which would cause undefined behavior upon drop as a result of calling
+// the wrong JNI function to delete the reference.
```

Comment:
Instead of having this documented special-case consideration that we need to be careful to not forget I wonder if we should look at adding an `IsLocalRef` trait (or similar) that can be used like a label/tag for all the types that can be safely wrapped via `AutoLocal`.

We can consider this in a follow up issue / PR.

Reply:
I don't think that's possible. It's unsafe to wrap `JObject<'static>` in `AutoLocal`, but it's *not* unsafe to wrap `JObject` with any other lifetime in `AutoLocal`, and I don't think there's any way to implement a trait for all lifetimes *except* `'static`.

We could add another wrapper type, `Local<'local, T>`, where `T` is `JObject` or something that wraps it, and `'local` is the local reference frame it was created in. Then, `JNIEnv::auto_local` accepts `Local<'local, T>` and returns `AutoLocal<'local, T>`. `JObject` and the other object wrapper types would no longer have a lifetime parameter and would become borrowed-only types like `str`, never appearing in owned form.

This would also make it possible to generify `GlobalRef` and `WeakRef` without using a GAT to change the lifetime.

This would be quite the footgun, though, since every `extern fn` that takes a parameter of type `JObject`/`JClass`/etc has to be changed to take `Local<JObject>`/`Local<JClass>`/etc, and if you forget any, you get undefined behavior! Rust has a way to solve this problem, [`extern type`](https://github.com/rust-lang/rfcs/blob/master/text/1861-extern-types.md), but [it's unstable and has been for a long time](https://github.com/rust-lang/rust/issues/43467).